e2e-testing.md

```diff
@@ -0,0 +1,321 @@
+  }
+
+  async linkToCourse(
+    courseId: number,
+    status: CourseUserStatus,
+  ): Promise<void> {
+    await this.api.seedItemIfNotExist(
+      FactoryItemType.CourseUsers,
+      {
+        courseId,
+        userId: this.id,
+        status,
+      },
+    );
+  }
+}
+```
+
+2\. Page objects
+--------------
+
+#### 2.1. Define page locators outside the constructor
+
+Please define locators in the page object classes and components using the below example.
+There could be different approaches, but we agreed to follow a single style within the organization.
+
+```typescript
+// ❌ bad
+export class SignInPage extends BasePage{
+  private readonly emailField: Locator;
+
+  constructor(page: Page) {
+      super(page);
+      this.emailField = page.getByTestId('sign-in-user-email');
+  }
+}
+
+// ✅ good
+export class SignInPage extends BasePage{
+  private readonly emailField = this.page.getByTestId('sign-in-user-email');
+
+  constructor(page: Page) {
+    super(page);
+  }
+}
+```
+
+
+
+#### 2.2. Parametrized page URLs defining
+
+Use the below approach to define the parametrized page object URLs.
+
+```typescript
+// ✅ good
+interface Options {
+  chatId: number;
+}
+
+export class ChatPage extends LoggedInBasePage {
+
+  constructor(
+    page: Page,
+    options: Options,
+  ) {
+    super(page);
+
+    this.url = ROUTES.chat(options.chatId).index;
+  }
+}
+```
+
+
+#### 2.3. Use Components to re-use common page element groups
+
+
+If you defined some common elements from the header, footer, sideBar, or some popups that appear on several pages - make sure to define them in the corresponding component class.
+
+
+
+#### 2.4. Do not use dynamic values in allure step text
+
+
+Do not use dynamic values in allure step text becuase that way new test case will be added to the Allure TestOps on each test re-run.
+
+```typescript
+// ❌ bad
+async typeCourseName(name: string): Promise<void> {
+  await test.step(`Type course name ${name}`, async () => {
+    await this.courseDropwDown.type(name);
+  });
+}
+
+
+// ✅ good
+async typeCourseName(name: string): Promise<void> {
+  await test.step('Type course name', async () => {
+    await this.courseDropwDown.type(name);
+  });
+}
+
+```
+
+
+3\. Test scenarios
+--------------
+
+#### 3.1. Use snake_case for test folders naming
+
+
+Use the snake_case for test folder names. Never use spaces in the folders and file names.
+
+
+```typescript
+// ❌ bad
+- tests
+    - e2e
+        - Admin Tools
+            - Homework-Review-Plugin
+
+
+// ✅ good
+- tests
+   - e2e
+        - Admin_Tools
+            - Homework_Review_Plugin
+
+```
+
+
+#### 3.2. Use folders for test suite organization
+
+
+Do not add test spec files from different suites to one folder. Create separate folder for each test suite.
+
+
+#### 3.3. Use one spec file per one test
+
+
+Do not add several tests into one spec file. Each spec file should normally have one test scenario. The exception can be done for the parametrized tests.
```

Comment:
could you also add examples?

like 
bad:

- test if page opened
- test if registration form works
- test if user is redirected after registration to new application
- test if new application form can be submitted

good:
- test FT application scenario: open page, fill registration form, fill application form, make sure it's submitted